diff --git a/dbus/gattlib_adapter.c b/dbus/gattlib_adapter.c
index d3d4b61..550fbaa 100644
--- a/dbus/gattlib_adapter.c
+++ b/dbus/gattlib_adapter.c
@@ -273,7 +273,7 @@ on_interface_proxy_properties_changed (GDBusObjectManagerClient *device_manager,
 
 		enum _gattlib_device_state old_device_state = gattlib_device_get_state(gattlib_adapter, proxy_object_path);
 
-		if (old_device_state == NOT_FOUND) {
+		if (old_device_state == NOT_FOUND || old_device_state == DISCONNECTED) {
 			if (has_rssi || has_manufacturer_data) {
 				int ret = gattlib_device_set_state(gattlib_adapter, proxy_object_path, DISCONNECTED);
 				if (ret == GATTLIB_SUCCESS) {
